zipline/finance/asset_restrictions.py

```diff
@@ -0,0 +1,215 @@
+import abc
+from numpy import vectorize
+from functools import partial, reduce
+import operator
+import pandas as pd
+from six import with_metaclass
+from collections import namedtuple
+from itertools import groupby
+
+from zipline.utils.enum import enum
+from zipline.utils.numpy_utils import vectorized_is_element
+from zipline.assets import Asset
+
+
+Restriction = namedtuple(
+    'Restriction', ['asset', 'effective_date', 'state']
+)
+
+
+RESTRICTION_STATES = enum(
+    'ALLOWED',
+    'FROZEN',
+)
+
+
+class Restrictions(with_metaclass(abc.ABCMeta)):
+    """
+    Abstract restricted list interface
+    """
+
+    @abc.abstractmethod
+    def is_restricted(self, assets, dt):
+        """
+        Is the asset restricted (RestrictionStates.FROZEN) on the given dt?
+
+        Parameters
+        ----------
+        asset : Asset of iterable of Assets
+            The asset(s) for which we are querying a restriction
+        dt : pd.Timestamp
+            The timestamp of the restriction query
+
+        Returns
+        -------
+        is_restricted : bool or pd.Series[bool] indexed by asset
+            Is the asset or assets restricted on this dt?
+
+        """
+        raise NotImplementedError('is_restricted')
+
+    def __or__(self, other_restriction):
+        """
+        Base implementation for combining two restrictions. If the right side
+        is a _UnionRestrictions, calls the overriding implementation with
+        _UnionRestrictions on the left side
+        """
+        if isinstance(other_restriction, _UnionRestrictions):
+            return _UnionRestrictions.__or__(other_restriction, self)
+        return _UnionRestrictions([self, other_restriction])
+
+
+class _UnionRestrictions(Restrictions):
+    """
+    A union of a number of sub restrictions
+
+    Parameters
+    ----------
+    sub_restrictions : iterable of Restrictions (but not _UnionRestrictions)
+        The Restrictions to be added together
+    """
+
+    def __new__(cls, sub_restrictions):
+        """
```

Comment:
Generally we don't put docstrings on `__init__` and `__new__`, since the constructor for a type gets documented in the class docstring.